### DISCUSSION TITLE
Add Macros support on Accounts struct (Currently IDL generator doesn't seem to support them)

### DISCUSSION THREAD

**Discussion by AnderUstarroz** [otter-sec/anchor#2432] [Ideas]
I have created an [issue with the same name](https://github.com/coral-xyz/anchor/issues/2431), but I wanted to raise the problem here to see what is your take on this. 

PD: Sorry for repeating myself

### Problem
Anchor IDL generator doesn't seem to support  macros on `Accounts` structs.

### Why is this important
Supporting macros on `Accounts` structs would facilitate the integration of 3rd party packages by abstracting away the complexity and reducing a lot the boilerplate. This is easier to understand with an example.

#### Example: Analytics package
Imagine we want to create an Analytics package which stores data on every instruction. We would have to repeat this boilerplate on every single instruction:
 
```rust
#[derive(Accounts)]
pub struct DoSomething<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,
    #[account(mut, seeds = [b"some_account".as_ref(), some_account.key().as_ref()],  bump = some_account.bump)],
    pub some_account:  Box<Account<'info, MyAccount>>,
    #[account(
        mut,
        seeds = [b"analytics-app".as_ref(), analytics_app.key().as_ref()], 
        bump = analytics_app.bump
    )]
    pub analytics_app:  Box<Account<'info, AnalyticsApp>>,
    #[account(
        mut,
        seeds = [b"analytics-owner".as_ref(), analytics_owner.key().as_ref()], 
        bump = analytics_owner.bump
    )]
    pub analytics_owner:  Box<Account<'info, AnalyticsOwner>>,
    pub analytics_program: Program<'info, Analytics>,
}
```

But using a simple Macro like syntax `#[analytics_accounts]`, all the analytics accounts could be added during compilation time making for a much succinct and neat integrations:

```rust
#[analytics_accounts] <--- All analytics accounts are added on compilation time. 
#[derive(Accounts)]
pub struct DoSomething<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,
    #[account(mut, seeds = [b"some_account".as_ref(), some_account.key().as_ref()],  bump = some_account.bump)],
    pub some_account:  Box<Account<'info, MyAccount>>,
}
```

## Benefits
Analytics, Authentication, Logs, Audits systems are just some examples of potential packages which would benefit from a "more mature" IDL generator in Anchor. I think this will add flexibility, increase code quality and facilitate the integration of packages by abstracting away complexity and avoiding having to write so much redundant code.

**Comment by peterschwarzdev** [comment]
I am surprised that there is no feedback from the maintainers, I think your point is very solid. I have faced the  same exact problem, which highlights the limitations of how Anchor works at the moment. IDL generation should definitely be improved so it works when adding accounts via Rust macros. Feels kind of wrong that Anchor makes extensive use of Macros while at the same time doesn't seem to allow (or support) developers using them.

**Comment by Aursen** [comment]
I want to take it into consideration but most of the Anchor maintainers work for free on their free time. So it's complicated to be everywhere but I will try to do some things. Unfortunately I don't have access to rest to sort out the priority list

**Reply by peterschwarzdev** [reply]
> I want to take it into consideration but most of the Anchor maintainers work for free on their free time. So it's complicated to be everywhere but I will try to do some things. Unfortunately I don't have access to rest to sort out the priority list

Wow, I didn't know that. I though such a critical tool would be properly funded by either [Solana](https://solana.com/) or [Metaplex](https://www.metaplex.com/).

**Comment by AnderUstarroz** [comment]
Just for the record @kklas already created an interesting [PR](https://github.com/coral-xyz/anchor/pull/2011) which seems to be addressing this issue (among others).
